src/mmfatl.h

```diff
@@ -0,0 +1,231 @@
+/*****************************************************************************/
+/*            Copyright (C) 2022  Wolf Lammen                                */
+/*            License terms:  GNU General Public License                     */
+/*****************************************************************************/
+/*34567890123456 (79-character line to adjust editor window) 2345678901234567*/
+
+#ifndef METAMATH_MMFATL_H_
+#define METAMATH_MMFATL_H_
+
+/*!
+ * \file mmfatl.h
+ * \brief supports generating of fatal error messages
+ *
+ * part of the application's infrastructure
+ *
+ * Rationale
+ * =========
+ *
+ * When a fatal error occurs, the internal structures of a program may be
+ * corrupted to the point that recovery is impossible.  The program exits
+ * immediately, but hopefully still displays a diagnostic message.
+ *
+ * To display this final message, we restrict its code to very basic,
+ * self-contained routines, independent of the rest of the program to the
+ * extent possible, thus avoiding any corrupted data.
+ *
+ * In particular everything should be pre-allocated and initialized, so the
+ * risk of a failure in a corrupted or memory-tight environment is minimized.
+ * This is to the detriment of flexibility, in particular, support for dynamic
+ * behavior is limited.
+ *
+ * A corrupt state is often caused by limit violations overwriting adjacent
+ * memory.  To specifically guard against this, the pre-allocated memory area,
+ * at your option, may include safety borders detaching necessary pre-set
+ * administrative data from other memory.
+ *
+ * Often it is sensible to embed details in a diagnosis message.  Placeholders
+ * in the format string mark insertion points for such values, much as in
+ * \p printf. The variety and functionality is greatly reduced in our case,
+ * though.  Only pieces of text or unsigned integers can be embedded
+ * (%s or %u placeholder).
+ *
+ * For this kind of expansion you still need a buffer where the final message
+ * is constructed.  In our context, this buffer is pre-allocated, and fixed in
+ * size, truncation of overflowing text enforced.
+ *
+ * Documentation
+ * =============
+ *
+ * Most comments are written in <a href="https://doxygen.nl/index.html">
+ * doxygen</a> (Qt variant) style.  If you have doxygen installed on your
+ * computer, you may generate HTML documentation out of them with its root
+ * page placed in build/html/index.html by running build.sh with the -d option.
```

Comment:
Why is this here? Is the general documentation style related to fatal errors in some way?

Reply:
To my knowledge no general documentation generation is installed throughout Metamath.  I added some support of doxygen a couple of months ago, but that is by no means finished.  This is contrary to the situation here: Both the interface here and its implementation in mmfatl.c is documented.  Once this holds for all other mm* files, AND a common style is agreed upon, I am happy to move this into a central location.

Reply:
> To my knowledge no general documentation generation is installed throughout Metamath. I added some support of doxygen a couple of months ago, but that is by no means finished. This is contrary to the situation here: Both the interface here and its implementation in mmfatl.c is documented. Once this holds for all other mm* files, AND a common style is agreed upon, I am happy to move this into a central location.

We [already discussed this before](https://github.com/metamath/metamath-exe/issues/33), and we agreed to use Doxygen, so there isn't any reason to hedge about this. If you want to document that you are using doxygen, it should be in the README.md or CONTRIBUTING.md anyway. It's entirely unnecessary to document that the code is documented in the documentation -- we can see that.

> Further note that this header file exports 4 simple routines dedicated to pre-allocation and deallocation of memory needed for formatting under extreme conditions, one formatter (like printf) and one exit call (close to abort). That's all. The criticized complexity of the implementation is hidden behind this interface, and is not exposed. Feel free to replace it with a handful of printf and abort lines.

Can you send me a PR with this? That sounds a lot easier to merge, and it's no worse than what the code is already doing (indeed, it fixes the main identified actual bug, which is that `outOfMemory()` is re-entrant).

Reply:
The hint towards doxygen should not be part of a doxygen comment.  This should be corrected.